utils/smart_resource.py

```diff
@@ -0,0 +1,301 @@
+            " ".join(
+                "@{}={}".format(
+                    key,
+                    repr(getattr(self, key)) if getattr(self, key) is not Unknown else "?")
+                for key in self.__fields__.iterkeys()
+            ),
+            str([key for key in self.__fields__.iterkeys() if key in self.__updates__])
+        )
+
+    @property
+    def needs_update(self):
+        return len(self.__updates__.keys()) > 0
+
+    @property
+    def needs_pull(self):
+        return any(value is Unknown for value in self.__data__.itervalues())
+
+    @property
+    def _underscore_object_name(self):
+        s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', self.__class__.__name__)
+        return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+    @property
+    def _nav_context(self):
+        return {self._underscore_object_name: self}
+
+    def _pull(self):
+        detail_fields = {}
+        edit_fields = {}
+        for field_name, field in self.__fields__.iteritems():
+            if field.GET_FROM_DETAIL:
+                detail_fields[field_name] = field
+            else:
+                edit_fields[field_name] = field
+        if detail_fields:
+            sel.force_navigate(self.DETAIL_LOCATION, context=self._nav_context)
+            for field_name, field in detail_fields.iteritems():
+                try:
+                    self.__data__[field_name] = field.get_value()
+                except AttributeError:
+                    self.__data__[field_name] = Impossible
+        if edit_fields:
+            sel.force_navigate(self.EDIT_LOCATION, context=self._nav_context)
+            for field_name, field in edit_fields.iteritems():
+                try:
+                    self.__data__[field_name] = field.get_value()
+                except AttributeError:
+                    self.__data__[field_name] = Impossible
+
+    def _fill(self, only_updates):
+        for field_name, field in self.__fields__.iteritems():
+            if not field.can_be_set:
+                continue
+            value = self.__updates__.get(
+                field_name,
+                None if only_updates else self.__data__.get(field_name)
+            )
+            if value is not None:
+                field.set_value(value)
+                if field_name in self.__updates__:
+                    self.__data__[field_name] = value
+                    self.__updates__.pop(field_name)
+
+    def go_to_create(self):
+        if self.CREATE_LOCATION is None:
+            raise Exception("Cannot create such type of resource!")
+        sel.force_navigate(self.CREATE_LOCATION)
+
+    def go_to_update(self):
+        if self.EDIT_LOCATION is None:
+            raise Exception("Cannot update such type of resource!")
+        sel.force_navigate(self.EDIT_LOCATION, context=self._nav_context)
+
+    def go_to_detail(self):
+        sel.force_navigate(self.DETAIL_LOCATION, context=self._nav_context)
+
+    def create(self):
```

Comment:
Thanks for comments :) will be considered after I wake up properly after two weeks of doing nothing :smile: